Rules/UseUsingScopeModifierInNewRunspaces.cs

```diff
@@ -0,0 +1,375 @@
+
+                var varsInAssignments = FindVarsInAssignmentAsts(ast);
+                
+                nonAssignedNonUsingVars.AddRange(
+                    FindNonAssignedNonUsingVarAsts(ast, varsInAssignments));
+            }
+            
+            foreach (var variableExpression in nonAssignedNonUsingVars)
+            {
+                if (variableExpression == null)
+                    continue;
+
+                yield return new DiagnosticRecord(
+                    message: string.Format(CultureInfo.CurrentCulture,
+                        Strings.UseUsingScopeModifierInNewRunspacesError, variableExpression.ToString()),
+                    extent: variableExpression.Extent,
+                    ruleName: GetName(),
+                    severity: DiagnosticSeverity.Warning,
+                    scriptPath: fileName,
+                    ruleId: variableExpression.ToString(),
+                    suggestedCorrections: GetSuggestedCorrections(ast: variableExpression));
+            }
+        }
+
+        private static IEnumerable<CorrectionExtent> GetSuggestedCorrections(Ast ast)
+        {
+            var varWithUsing = "$using:" + ast.Extent.Text.TrimStart('$');
+            var description = string.Format(
+                CultureInfo.CurrentCulture,
+                Strings.UseUsingScopeModifierInNewRunspacesCorrectionDescription,
+                ast.Extent.Text,
+                varWithUsing);
+            var corrections = new List<CorrectionExtent>()
+            {
+                new CorrectionExtent(
+                    startLineNumber: ast.Extent.StartLineNumber,
+                    endLineNumber: ast.Extent.EndLineNumber,
+                    startColumnNumber: ast.Extent.StartColumnNumber,
+                    endColumnNumber: ast.Extent.EndColumnNumber,
+                    text: varWithUsing,
+                    file: ast.Extent.File,
+                    description: description
+                )
+            };
+
+            return corrections;
+        }
+
+        private static IEnumerable<VariableExpressionAst> ProcessInvokeCommandSessionAsts(ScriptBlockAst scriptBlockAst)
+        {
+            var invokeCommandCmdletNamesAndAliases = Helper.Instance.CmdletNameAndAliases("Invoke-Command");
+            var sessionDictionary = new Dictionary<string, List<Ast>>();
+            var result = new List<VariableExpressionAst>();
+
+            // The shortest unambiguous name for parameter -Session is 'session' (SessionName and SessionOption) also exist.
+            var scriptBlockExpressionAstsToAnalyze = scriptBlockAst.FindAll(
+                predicate: a => a is ScriptBlockExpressionAst scriptBlockExpressionAst &&
+                                scriptBlockExpressionAst.Parent is CommandAst commandAst &&
+                                invokeCommandCmdletNamesAndAliases
+                                    .Contains(commandAst.GetCommandName(), StringComparer.OrdinalIgnoreCase) &&

```

Comment:
`commandAst.GetCommandName()` can be null, which will cause a crash here

Reply:
Example:

```powershell
{ & $commandName }.Ast.EndBlock.Statements[0].PipelineElements[0].GetCommandName()
```

Reply:
Thanks, noted, I'm thinking of adding a test along these lines to be sure.